### PR TITLE
feat: update table metadata in lock

### PR DESCRIPTION
**Pull request by MichaelScofield**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Using distributed lock to guard against the concurrent updating of table metadatas in region failover procedure.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

#1126

### CODE REVIEW

File: src/meta-srv/src/lock/memory.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+use async_trait::async_trait;
+use tokio::sync::{Mutex, OwnedMutexGuard};
+
+use crate::error::Result;
+use crate::lock::{DistLock, Key, Opts};
+
+#[derive(Default)]
+pub(crate) struct MemLock {
+    mutexes: Arc<Mutex<HashMap<Key, Arc<Mutex<()>>>>>,
```
Comment (Fengys123) [inline src/meta-srv/src/lock/memory.rs:26]: Looks like Arc can be removed.
```suggestion
    mutexes: Mutex<HashMap<Key, Arc<Mutex<()>>>>,
```

Comment (Fengys123) [inline src/meta-srv/src/lock/memory.rs:26]: NIT: dashmap is better?
```suggestion
    mutexes: Dashmap<Key, Arc<Mutex<()>>>,
```

**Review by Fengys123 — APPROVED** [review]
LGTM, and left some comments

File: src/meta-srv/src/lock/memory.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#[derive(Default)]
+pub(crate) struct MemLock {
+    mutexes: DashMap<Key, Arc<Mutex<()>>>,
+    guards: DashMap<Key, OwnedMutexGuard<()>>,
+}
+
+#[async_trait]
+impl DistLock for MemLock {
+    async fn lock(&self, name: Vec<u8>, _opts: Opts) -> Result<Key> {
+        let key = name;
```
Comment (killme2008) [inline src/meta-srv/src/lock/memory.rs:33]: The alias is not necessary, just rename the argument to `key`.

File: src/meta-srv/src/procedure/region_failover/update_metadata.rs [hunk 3]
```diff
@@ -183,19 +200,29 @@ impl State for UpdateRegionMetadata {
...
+                        failed_region, e
+                    ),
+                }
+                .build()
+            })?;
         Ok(Box::new(RegionFailoverEnd))
     }
 }
 
+fn table_metadata_lock_key(region: &RegionIdent) -> Key {
```
Comment (killme2008) [inline src/meta-srv/src/procedure/region_failover/update_metadata.rs:218]: Can we move all these distributed locking key generate functions into a single mod? It's much more readable and maintainable. The maintainer can read all the keys in one place.

**Comment by MichaelScofield** [comment]
@killme2008 PTAL

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**